src/server/http/http.go

```diff
@@ -0,0 +1,172 @@
+	"net/http"
+	"net/http/httputil"
+	"net/url"
+	"path"
+	"strings"
+	"sync"
+
+	"github.com/pachyderm/pachyderm/src/client"
+	"github.com/pachyderm/pachyderm/src/client/auth"
+
+	"github.com/gogo/protobuf/types"
+	"github.com/julienschmidt/httprouter"
+	"golang.org/x/net/context"
+	"google.golang.org/grpc/metadata"
+)
+
+// HTTPPort specifies the port the server will listen on
+const HTTPPort = 652
+const apiVersion = "v1"
+
+func versionPath(p string) string {
+	return path.Join("/", apiVersion, p)
+}
+
+var (
+	getFilePath = versionPath("pfs/repos/:repoName/commits/:commitID/files/*filePath")
+	servicePath = versionPath("pps/services/:serviceName/*path")
+	loginPath   = versionPath("auth/login")
+	logoutPath  = versionPath("auth/logout")
+)
+
+type router = *httprouter.Router
+
+type server struct {
+	router
+	address        string
+	pachClient     *client.APIClient
+	pachClientOnce sync.Once
+	httpClient     *http.Client
+}
+
+// NewHTTPServer returns a Pachyderm HTTP server.
+func NewHTTPServer(address string) (http.Handler, error) {
+	router := httprouter.New()
+	s := &server{
+		router:     router,
+		address:    address,
+		httpClient: &http.Client{},
+	}
+
+	router.GET(getFilePath, s.getFileHandler)
+	router.POST(loginPath, s.authLoginHandler)
+	router.POST(logoutPath, s.authLogoutHandler)
+	router.GET(servicePath, s.serviceHandler)
+	router.POST(servicePath, s.serviceHandler)
+	// Debug method (to check login cookies):
+	router.GET(loginPath, s.loginForm)
+	router.NotFound = http.HandlerFunc(notFound)
+	return s, nil
+}
+
+func (s *server) getFileHandler(w http.ResponseWriter, r *http.Request, ps httprouter.Params) {
+	filePaths := strings.Split(ps.ByName("filePath"), "/")
+	fileName := filePaths[len(filePaths)-1]
+	ctx := context.Background()
+	for _, cookie := range r.Cookies() {
+		if cookie.Name == auth.ContextTokenKey {
+			ctx = metadata.NewIncomingContext(
+				ctx,
+				metadata.Pairs(auth.ContextTokenKey, cookie.Value),
+			)
+		}
+	}
+	downloadValues := r.URL.Query()["download"]
+	if len(downloadValues) == 1 && downloadValues[0] == "true" {
+		w.Header().Add("Content-Disposition", fmt.Sprintf("attachment; filename=\"%v\"", fileName))
+	}
+	c := s.getPachClient()
+	commitInfo, err := c.InspectCommit(ps.ByName("repoName"), ps.ByName("commitID"))
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+	content, err := c.GetFileReadSeeker(ps.ByName("repoName"), ps.ByName("commitID"), ps.ByName("filePath"))
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+	modtime, err := types.TimestampFromProto(commitInfo.Finished)
+	http.ServeContent(w, r, fileName, modtime, content)
+	return
+}
+
+func (s *server) serviceHandler(w http.ResponseWriter, r *http.Request, ps httprouter.Params) {
+	c := s.getPachClient()
+	pipelineInfo, err := c.InspectPipeline(ps.ByName("serviceName"))
+	if err != nil {
+		// TODO this could be a NotFound which should 404, not 500
```

Comment:
That sounds like the right thing to do here. Can we add that in? There's the `notFound(...)` helper below.

Reply:
Yup, I'll do that.